libs/core/algorithms/include/hpx/parallel/unseq/simd_helpers.hpp

```diff
@@ -0,0 +1,161 @@
+        Compiler and Hardware should also support vector operations for IterDiff,
+        else we see slower performance when compared to sequential version
+    */
+    template <class Iter, class IterDiff, class F>
+    Iter unseq_first_n(const Iter first, const IterDiff n, F&& f_) noexcept
+    {
+        /*
+            OMP loops can not have ++Iter, only integral types are allowed
+            Hence perform arthemetic on Iterators
+            which is O(1) only in case of random access iterators
+        */
+        static_assert(hpx::traits::is_random_access_iterator_v<Iter>,
+            "algorithm is efficient only in case of Random Access Iterator");
+#if HPX_EARLYEXIT_PRESENT
+        IterDiff i = 0;
+        // clang-format off
+        HPX_PRAGMA_VECTOR_UNALIGNED HPX_PRAGMA_SIMD_EARLYEXIT
+        for (; i < n; ++i)
+        {
+            if (HPX_INVOKE(f, first, i))
+            {
+                break;
+            }
+        }
+        // clang-format on
+
+        return first + i;
+#else
+        auto f = [](auto it, int ind) { return HPX_INVOKE(f_, *(ind + it)) };
+        // int32_t has best support for vectorization from compilers and hardware
+        IterDiff i = 0;
+        static constexpr int32_t numBlocks = HPX_LANE_SIZE / sizeof(int32_t);
+        alignas(HPX_LANE_SIZE) int32_t simdLane[numBlocks] = {0};
+        while (i <= n - numBlocks)
+        {
+            int32_t foundFlag = 0;
+
+            // clang-format off
+            HPX_PRAGMA_VECTOR_UNALIGNED HPX_VECTOR_REDUCTION(| : foundFlag)
+            for (IterDiff j = i; j < i + numBlocks; ++j)
+            {
+                const int32_t t = HPX_INVOKE(f,first, j);
+                simdLane[j - i] = t;
+                foundFlag |= t;
+            }
+            // clang-format on
+
+            if (foundFlag)
+            {
+                IterDiff j;
+                for (j = 0; j < numBlocks; ++j)
+                {
+                    if (simdLane[j])
+                    {
+                        break;
+                    }
+                }
+                return first + i + j;
+            }
+            i += numBlocks;
+        }
+
+        //Keep remainder scalar
+        while (i != n)
+        {
+            if (HPX_INVOKE(f, first, i))
+            {
+                return first + i;
+            }
+            ++i;
+        }
+        return first + i;
+#endif    //HPX_EARLYEXIT_PRESENT
+    }
+
+    template <class Iter1, class Iter2, class IterDiff, class F>
+    std::pair<Iter1, Iter2> unseq2_first_n(const Iter1 first1,
+        const Iter2 first2, const IterDiff n, F&& f) noexcept
+    {
+#if HPX_EARLYEXIT_PRESENT
+        IterDiff i = 0;
+
+        // clang-format off
+        HPX_PRAGMA_VECTOR_UNALIGNED HPX_PRAGMA_SIMD_EARLYEXIT
+        for (; i < n; ++i)
+            if (HPX_INVOKE(f, *(first1 + i), *(first2 + i)))
+                break;
+        // clang-format on
+
+        return std::make_pair(first1 + i, first2 + i);
```

Comment:
```suggestion
        HPX_PRAGMA_VECTOR_UNALIGNED HPX_PRAGMA_SIMD_EARLYEXIT
        for (IterDiff i = 0; i < n; ++i)
        {
            if (HPX_INVOKE(f, *first1, *first2))
                break;
            ++first1; ++first2;
        }
        // clang-format on
        return std::make_pair(first, first2);
```
Or would this break vectorization?

Reply:
The block of code is only called when the program is compiled with an intel compiler of version >= 1800.
I haven't been able to test out the code on an intel compiler but given how OMP works, it'd very likely throw a compile time error.